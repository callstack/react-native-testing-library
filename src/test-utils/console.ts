import { format } from 'util';

export function excludeConsoleMessage(logFn: (...args: unknown[]) => void, excludeMessage: string) {
  return (...args: unknown[]) => {
    const message = format(...args);
    if (message.includes(excludeMessage)) {
      return;
    }

    logFn(...args);
  };
}
