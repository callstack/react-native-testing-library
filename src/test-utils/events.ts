interface EventEntry {
  name: string;
  payload: any;
}

export function createEventLogger() {
  const events: EventEntry[] = [];
  const logEvent = (name: string) => {
    return (event: unknown) => {
      const eventEntry: EventEntry = {
        name,
        payload: event,
      };

      events.push(eventEntry);
    };
  };

  return { events, logEvent };
}
