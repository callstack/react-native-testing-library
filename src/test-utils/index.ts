export * from './console';
export * from './events';
