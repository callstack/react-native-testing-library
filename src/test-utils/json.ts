import type { ReactTestRendererJSON } from 'react-test-renderer';

type JsonPropsMapper = {
  [key: string]: unknown;
};

export function mapJsonProps<T extends ReactTestRendererJSON | ReactTestRendererJSON[] | null>(
  element: T,
  mapper: JsonPropsMapper,
): T {
  if (Array.isArray(element)) {
    return element.map((e) => mapJsonProps(e, mapper)) as T;
  }

  if (!element) {
    return element;
  }

  const resultProps = { ...element.props };
  Object.keys(mapper).forEach((key) => {
    if (key in element.props) {
      resultProps[key] = mapper[key];
    }
  });

  const resultElement = { ...element, props: resultProps };
  Object.defineProperty(resultElement, '$$typeof', {
    value: Symbol.for('react.test.json'),
  });

  return resultElement;
}
