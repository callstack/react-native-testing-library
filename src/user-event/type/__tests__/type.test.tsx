import * as React from 'react';
import { TextInput, TextInputProps, View } from 'react-native';
import { createEventLogger } from '../../../test-utils';
import { render, screen } from '../../..';
import { userEvent } from '../..';

beforeEach(() => {
  jest.useRealTimers();
});

function renderTextInputWithToolkit(props: TextInputProps = {}) {
  const { events, logEvent } = createEventLogger();

  render(
    <TextInput
      testID="input"
      onFocus={logEvent('focus')}
      onBlur={logEvent('blur')}
      onPressIn={logEvent('pressIn')}
      onPressOut={logEvent('pressOut')}
      onChange={logEvent('change')}
      onChangeText={logEvent('changeText')}
      onKeyPress={logEvent('keyPress')}
      onTextInput={logEvent('textInput')}
      onSelectionChange={logEvent('selectionChange')}
      onSubmitEditing={logEvent('submitEditing')}
      onEndEditing={logEvent('endEditing')}
      onContentSizeChange={logEvent('contentSizeChange')}
      {...props}
    />
  );

  return {
    ...screen,
    events,
  };
}

describe('type()', () => {
  it('supports basic case', async () => {
    jest.spyOn(Date, 'now').mockImplementation(() => 100100100100);
    const { events, ...queries } = renderTextInputWithToolkit();

    const user = userEvent.setup();
    await user.type(queries.getByTestId('input'), 'abc');

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual([
      'pressIn',
      'focus',
      'pressOut',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'endEditing',
      'blur',
    ]);

    expect(events).toMatchSnapshot('input: "abc"');
  });

  it.each(['modern', 'legacy'])('works with %s fake timers', async (type) => {
    jest.useFakeTimers({ legacyFakeTimers: type === 'legacy' });
    const { events, ...queries } = renderTextInputWithToolkit();

    const user = userEvent.setup();
    await user.type(queries.getByTestId('input'), 'abc');

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual([
      'pressIn',
      'focus',
      'pressOut',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'endEditing',
      'blur',
    ]);
  });

  it('supports defaultValue prop', async () => {
    const { events, ...queries } = renderTextInputWithToolkit({
      defaultValue: 'xxx',
    });

    const user = userEvent.setup();
    await user.type(queries.getByTestId('input'), 'ab');

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual([
      'pressIn',
      'focus',
      'pressOut',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'endEditing',
      'blur',
    ]);

    expect(events).toMatchSnapshot('input: "ab", defaultValue: "xxx"');
  });

  it('does respect editable prop', async () => {
    const { events, ...queries } = renderTextInputWithToolkit({
      editable: false,
    });

    const user = userEvent.setup();
    await user.type(queries.getByTestId('input'), 'ab');

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual([]);
  });

  it('supports backspace', async () => {
    const { events, ...queries } = renderTextInputWithToolkit({
      defaultValue: 'xxx',
    });

    const user = userEvent.setup();
    await user.type(queries.getByTestId('input'), '{Backspace}a');

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual([
      'pressIn',
      'focus',
      'pressOut',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'endEditing',
      'blur',
    ]);

    expect(events).toMatchSnapshot('input: "{Backspace}a", defaultValue: "xxx"');
  });

  it('supports multiline', async () => {
    const { events, ...queries } = renderTextInputWithToolkit({
      multiline: true,
    });

    const user = userEvent.setup();
    await user.type(queries.getByTestId('input'), '{Enter}\n');

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual([
      'pressIn',
      'focus',
      'pressOut',
      'keyPress',
      'textInput',
      'change',
      'changeText',
      'selectionChange',
      'contentSizeChange',
      'keyPress',
      'textInput',
      'change',
      'changeText',
      'selectionChange',
      'contentSizeChange',
      'endEditing',
      'blur',
    ]);

    expect(events).toMatchSnapshot('input: "{Enter}\\n", multiline: true');
  });

  test('skips press events when `skipPress: true`', async () => {
    const { events, ...queries } = renderTextInputWithToolkit();

    const user = userEvent.setup();
    await user.type(queries.getByTestId('input'), 'a', {
      skipPress: true,
    });

    const eventNames = events.map((e) => e.name);
    expect(eventNames).not.toContainEqual('pressIn');
    expect(eventNames).not.toContainEqual('pressOut');
    expect(eventNames).toEqual([
      'focus',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'endEditing',
      'blur',
    ]);
  });

  it('triggers submit event with `submitEditing: true`', async () => {
    const { events, ...queries } = renderTextInputWithToolkit();

    const user = userEvent.setup();
    await user.type(queries.getByTestId('input'), 'a', {
      submitEditing: true,
    });

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual([
      'pressIn',
      'focus',
      'pressOut',
      'keyPress',
      'change',
      'changeText',
      'selectionChange',
      'submitEditing',
      'endEditing',
      'blur',
    ]);

    expect(events[7].name).toBe('submitEditing');
    expect(events[7].payload).toEqual({
      nativeEvent: { text: 'a', target: 0 },
      currentTarget: {},
      target: {},
    });
  });

  it('works when not all events have handlers', async () => {
    const { events, logEvent } = createEventLogger();
    render(
      <TextInput
        testID="input"
        onChangeText={logEvent('changeText')}
        onEndEditing={logEvent('endEditing')}
      />
    );

    const user = userEvent.setup();
    await user.type(screen.getByTestId('input'), 'abc');

    const eventNames = events.map((e) => e.name);
    expect(eventNames).toEqual(['changeText', 'changeText', 'changeText', 'endEditing']);

    expect(events).toMatchSnapshot('input: "abc"');
  });

  it('does NOT work on View', async () => {
    render(<View testID="input" />);

    const user = userEvent.setup();
    await expect(
      user.type(screen.getByTestId('input'), 'abc')
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"type() works only with host "TextInput" elements. Passed element has type "View"."`
    );
  });

  // View that ignores props type checking
  const AnyView = View as React.ComponentType<any>;

  it('does NOT bubble up', async () => {
    const parentHandler = jest.fn();
    render(
      <AnyView
        onChangeText={parentHandler}
        onChange={parentHandler}
        onKeyPress={parentHandler}
        onTextInput={parentHandler}
        onFocus={parentHandler}
        onBlur={parentHandler}
        onEndEditing={parentHandler}
        onPressIn={parentHandler}
        onPressOut={parentHandler}
      >
        <TextInput testID="input" />
      </AnyView>
    );

    const user = userEvent.setup();
    await user.type(screen.getByTestId('input'), 'abc');
    expect(parentHandler).not.toHaveBeenCalled();
  });

  it('supports direct access', async () => {
    const { events, logEvent } = createEventLogger();
    render(
      <TextInput
        testID="input"
        onChangeText={logEvent('changeText')}
        onFocus={logEvent('focus')}
        onBlur={logEvent('blur')}
      />
    );

    await userEvent.type(screen.getByTestId('input'), 'abc');

    const eventNames = events.map((event) => event.name);
    expect(eventNames).toEqual(['focus', 'changeText', 'changeText', 'changeText', 'blur']);
  });
});
