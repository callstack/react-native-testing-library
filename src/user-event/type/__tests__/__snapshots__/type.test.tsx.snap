// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`type() supports backspace: input: "{Backspace}a", defaultValue: "xxx" 1`] = `
[
  {
    "name": "pressIn",
    "payload": {
      "currentTarget": {
        "measure": [Function],
      },
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "changedTouches": [],
        "identifier": 0,
        "locationX": 0,
        "locationY": 0,
        "pageX": 0,
        "pageY": 0,
        "target": 0,
        "timestamp": 100100100100,
        "touches": [],
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "focus",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "pressOut",
    "payload": {
      "currentTarget": {
        "measure": [Function],
      },
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "changedTouches": [],
        "identifier": 0,
        "locationX": 0,
        "locationY": 0,
        "pageX": 0,
        "pageY": 0,
        "target": 0,
        "timestamp": 100100100100,
        "touches": [],
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "Backspace",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "xx",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "xx",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 2,
          "start": 2,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "a",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "xxa",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "xxa",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 3,
          "start": 3,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "endEditing",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
        "text": "xxa",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "blur",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
]
`;

exports[`type() supports basic case: input: "abc" 1`] = `
[
  {
    "name": "pressIn",
    "payload": {
      "currentTarget": {
        "measure": [Function],
      },
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "changedTouches": [],
        "identifier": 0,
        "locationX": 0,
        "locationY": 0,
        "pageX": 0,
        "pageY": 0,
        "target": 0,
        "timestamp": 100100100100,
        "touches": [],
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "focus",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "pressOut",
    "payload": {
      "currentTarget": {
        "measure": [Function],
      },
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "changedTouches": [],
        "identifier": 0,
        "locationX": 0,
        "locationY": 0,
        "pageX": 0,
        "pageY": 0,
        "target": 0,
        "timestamp": 100100100100,
        "touches": [],
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "a",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "a",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "a",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 1,
          "start": 1,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "b",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "ab",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "ab",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 2,
          "start": 2,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "c",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "abc",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "abc",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 3,
          "start": 3,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "endEditing",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
        "text": "abc",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "blur",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
]
`;

exports[`type() supports defaultValue prop: input: "ab", defaultValue: "xxx" 1`] = `
[
  {
    "name": "pressIn",
    "payload": {
      "currentTarget": {
        "measure": [Function],
      },
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "changedTouches": [],
        "identifier": 0,
        "locationX": 0,
        "locationY": 0,
        "pageX": 0,
        "pageY": 0,
        "target": 0,
        "timestamp": 100100100100,
        "touches": [],
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "focus",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "pressOut",
    "payload": {
      "currentTarget": {
        "measure": [Function],
      },
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "changedTouches": [],
        "identifier": 0,
        "locationX": 0,
        "locationY": 0,
        "pageX": 0,
        "pageY": 0,
        "target": 0,
        "timestamp": 100100100100,
        "touches": [],
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "a",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "xxxa",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "xxxa",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 4,
          "start": 4,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "b",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "xxxab",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "xxxab",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 5,
          "start": 5,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "endEditing",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
        "text": "xxxab",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "blur",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
]
`;

exports[`type() supports multiline: input: "{Enter}\\n", multiline: true 1`] = `
[
  {
    "name": "pressIn",
    "payload": {
      "currentTarget": {
        "measure": [Function],
      },
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "changedTouches": [],
        "identifier": 0,
        "locationX": 0,
        "locationY": 0,
        "pageX": 0,
        "pageY": 0,
        "target": 0,
        "timestamp": 100100100100,
        "touches": [],
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "focus",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "pressOut",
    "payload": {
      "currentTarget": {
        "measure": [Function],
      },
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "changedTouches": [],
        "identifier": 0,
        "locationX": 0,
        "locationY": 0,
        "pageX": 0,
        "pageY": 0,
        "target": 0,
        "timestamp": 100100100100,
        "touches": [],
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "Enter",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "textInput",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "previousText": "",
        "range": {
          "end": 1,
          "start": 1,
        },
        "target": 0,
        "text": "
",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "
",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "
",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 1,
          "start": 1,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "contentSizeChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "contentSize": {
          "height": 32,
          "width": 0,
        },
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "keyPress",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "key": "Enter",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "textInput",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "previousText": "
",
        "range": {
          "end": 2,
          "start": 2,
        },
        "target": 0,
        "text": "

",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "change",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "eventCount": 0,
        "target": 0,
        "text": "

",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "changeText",
    "payload": "

",
  },
  {
    "name": "selectionChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "selection": {
          "end": 2,
          "start": 2,
        },
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "contentSizeChange",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "contentSize": {
          "height": 48,
          "width": 0,
        },
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "endEditing",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
        "text": "

",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
  {
    "name": "blur",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
]
`;

exports[`type() works when not all events have handlers: input: "abc" 1`] = `
[
  {
    "name": "changeText",
    "payload": "a",
  },
  {
    "name": "changeText",
    "payload": "ab",
  },
  {
    "name": "changeText",
    "payload": "abc",
  },
  {
    "name": "endEditing",
    "payload": {
      "currentTarget": {},
      "isDefaultPrevented": [Function],
      "isPersistent": [Function],
      "isPropagationStopped": [Function],
      "nativeEvent": {
        "target": 0,
        "text": "abc",
      },
      "persist": [Function],
      "preventDefault": [Function],
      "stopPropagation": [Function],
      "target": {},
      "timeStamp": 0,
    },
  },
]
`;
